File: backend/src/agent_workbench/services/topology_layout.py
"""Reshape BMS schematic layout from pack topology (mirrors frontend topologyLayout.ts)."""

from __future__ import annotations

import copy
import re
from typing import Any

TOPOLOGY_RE = re.compile(r"^(\d+)s(\d+)p$", re.IGNORECASE)
CELL_PIN_RE = re.compile(r"^CELL_", re.IGNORECASE)
NTC_PIN_RE = re.compile(r"^NTC_", re.IGNORECASE)


def _parse_topology(topology: str) -> tuple[int, int] | None:
    match = TOPOLOGY_RE.match(topology.strip())
    if not match:
        return None
    return int(match.group(1)), int(match.group(2))


def _cell_stack_dimensions(series: int, parallel: int) -> tuple[int, int]:
    width = min(720, max(140, 48 + series * 22))
    height = max(72, 56 + max(0, parallel - 1) * 36)
    return width, height


def _afe_dimensions(series: int) -> tuple[int, int]:
    width = min(300, max(170, 100 + series * 7))
    height = max(95, 80 + (series // 4) * 10)
    return width, height


def _series_cell_pin_label(series: int) -> str:
    return "CELL_1" if series <= 1 else f"CELL_1-{series}"


def _stack_label(series: int, parallel: int, chemistry: str) -> str:
    topo = f"{series}S{parallel}P" if parallel > 1 else f"{series}S"
    return f"{topo} {chemistry} Stack"


def _find_by_type(nodes: list[dict[str, Any]], node_type: str) -> dict[str, Any] | None:
    for node in nodes:
        if node.get("type") == node_type:
            return node
    return None


def _update_pin_label(node: dict[str, Any], pattern: re.Pattern[str], label: str) -> None:
    for pin in node.get("pins") or []:
        if isinstance(pin, dict) and pattern.search(str(pin.get("label", ""))):
            pin["label"] = label


def _shift_nodes_right(nodes: list[dict[str, Any]], from_x: float, delta_x: float) -> None:
    if delta_x <= 0:
        return
    for node in nodes:
        if float(node.get("x", 0)) >= from_x:
            node["x"] = float(node.get("x", 0)) + delta_x


def _enrich_node_label(node: dict[str, Any], components: dict[str, Any]) -> None:
    ref = node.get("component_ref")
    if not ref or ref not in components:
        return
    part = components[ref].get("part")
    if not isinstance(part, str) or not part.strip():
        return
    label = str(node.get("label", ""))
    if re.search(r"\s[–—-]\s", label):
        return
    node_type = str(node.get("type", ""))
    if node_type in {"cells", "bms_board", "contactor"}:
        return
    roles = {
        "mcu": "MCU",
        "cell_monitor_ic": "Cell Monitor AFE",
        "can_transceiver": "CAN Transceiver",
        "temperature_network": "Temp Network",
        "balancing": "Cell Balancing",
        "current_sensor": "Current Monitor",
        "memory": "EEPROM / Flash",
    }
    role = roles.get(node_type, node_type.replace("_", " "))
    node["label"] = f"{role} – {part}"


def _resolve_part_label(node: dict[str, Any], components: dict[str, Any]) -> None:
    _enrich_node_label(node, components)


def _adapt_pack_view(view: dict[str, Any], pack: dict[str, Any], series: int, parallel: int) -> None:
    nodes = view.get("nodes") or []
    cells = _find_by_type(nodes, "cells")
    if not cells:
        return

    prev_width = float(cells.get("width", 200))
    width, height = _cell_stack_dimensions(series, parallel)
    cells["width"] = width
    cells["height"] = height
    cells["label"] = _stack_label(series, parallel, str(pack.get("chemistry", "")))
    cells["y"] = max(280, 460 - height)

    width_delta = width - prev_width
    if width_delta > 0:
        _shift_nodes_right(nodes, float(cells.get("x", 0)) + prev_width + 20, width_delta)

    bms_board = _find_by_type(nodes, "bms_board")
    if bms_board:
        min_gap = 44
        bms_board["x"] = max(float(bms_board.get("x", 0)), float(cells.get("x", 0)) + width + min_gap)

    contactor = _find_by_type(nodes, "contactor")
    if contactor and bms_board:
        contactor["x"] = max(
            float(contactor.get("x", 0)),
            float(bms_board.get("x", 0)) + float(bms_board.get("width", 0)) + 36,
        )


def _inject_second_afe(
    nodes: list[dict[str, Any]],
    edges: list[dict[str, Any]],
    primary: dict[str, Any],
    series: int,
    part: str | None,
) -> None:
    if series <= 16 or any(n.get("id") == "cell_ic_2" for n in nodes):
        return

    second_series = series - 16
    start_cell = 17
    end_cell = start_cell + second_series - 1
    cell_label = f"CELL_{start_cell}" if second_series <= 1 else f"CELL_{start_cell}-{end_cell}"
    afe_w, afe_h = _afe_dimensions(second_series)

    nodes.append(
        {
            "id": "cell_ic_2",
            "type": "cell_monitor_ic",
            "label": f"{part} AFE (B)" if part else "Cell Monitor IC (B)",
            "x": primary.get("x", 60),
            "y": float(primary.get("y", 200)) + float(primary.get("height", 85)) + 36,
            "width": afe_w,
            "height": afe_h,
            "component_ref": f"{primary.get('component_ref')}_b" if primary.get("component_ref") else None,
            "telemetry": {"cells_monitored": second_series},
            "pins": [
                {"id": "cell_sense", "label": cell_label, "side": "left"},
                {"id": "i2c_sda", "label": "I2C_SDA", "side": "top"},
                {"id": "i2c_scl", "label": "I2C_SCL", "side": "top"},
                {"id": "bal_out", "label": "BAL_OUT", "side": "right"},
            ],
        }
    )

    mcu = _find_by_type(nodes, "mcu")
    if mcu:
        edges.append(
            {
                "from_node": mcu.get("id", "mcu"),
                "from_pin": "i2c_sda",
                "to_node": "cell_ic_2",
                "to_pin": "i2c_sda",
                "signal": "I2C_SDA",
            }
        )
        edges.append(
            {
                "from_node": mcu.get("id", "mcu"),
                "from_pin": "i2c_scl",
                "to_node": "cell_ic_2",
                "to_pin": "i2c_scl",
                "signal": "I2C_SCL",
            }
        )


def _adapt_bms_view(view: dict[str, Any], pack: dict[str, Any], series: int, components: dict[str, Any]) -> None:
    nodes = view.get("nodes") or []
    edges = view.get("edges") or []
    cell_label = _series_cell_pin_label(series)
    ntc_label = "NTC_1" if series <= 1 else f"NTC_1-{series}"

    cell_ic = _find_by_type(nodes, "cell_monitor_ic")
    if cell_ic:
        primary_series = 16 if series > 16 else series
        afe_w, afe_h = _afe_dimensions(primary_series)
        cell_ic["width"] = afe_w
        cell_ic["height"] = afe_h
        _update_pin_label(cell_ic, CELL_PIN_RE, _series_cell_pin_label(primary_series))
        telemetry = dict(cell_ic.get("telemetry") or {})
        telemetry["cells_monitored"] = primary_series
        cell_ic["telemetry"] = telemetry
        _resolve_part_label(cell_ic, components)

        part = None
        ref = cell_ic.get("component_ref")
        if ref and ref in components:
            raw = components[ref].get("part")
            part = raw if isinstance(raw, str) else None
            components[ref]["cells_monitored"] = primary_series if series > 16 else series
        if series > 16:
            _inject_second_afe(nodes, edges, cell_ic, series, part)
        elif ref and ref in components:
            components[ref]["cells_monitored"] = series

    temp_net = _find_by_type(nodes, "temperature_network")
    if temp_net:
        _update_pin_label(temp_net, NTC_PIN_RE, ntc_label)
        if cell_ic:
            temp_net["y"] = max(float(temp_net.get("y", 0)), float(cell_ic.get("y", 0)) + float(cell_ic.get("height", 0)) + 28)
        temp_telemetry = dict(temp_net.get("telemetry") or {})
        temp_telemetry["sensor_count"] = series
        temp_net["telemetry"] = temp_telemetry
        if "temp_network" in components:
            components["temp_network"]["sensor_count"] = series

    balancing = _find_by_type(nodes, "balancing")
    if balancing and cell_ic:
        balancing["y"] = max(float(balancing.get("y", 0)), float(cell_ic.get("y", 0)) + float(cell_ic.get("height", 0)) + 20)

    current = _find_by_type(nodes, "current_sensor")
    memory = _find_by_type(nodes, "memory")
    if current and memory:
        bottom_y = max(float(current.get("y", 0)), float(memory.get("y", 0)), 320)
        current["y"] = bottom_y
        memory["y"] = bottom_y + float(current.get("height", 70)) + 24
        if float(memory.get("x", 0)) <= float(current.get("x", 0)) + float(current.get("width", 0)):
            memory["x"] = float(current.get("x", 0)) + float(current.get("width", 140)) + 40

    for node in nodes:
        _resolve_part_label(node, components)


def reshape_architecture_for_topology(data: dict[str, Any]) -> dict[str, Any]:
    """Return a deep copy with pack-driven layout applied."""
    result = copy.deepcopy(data)
    pack = result.get("pack") or {}
    topology = str(pack.get("topology", ""))
    parsed = _parse_topology(topology)
    if not parsed:
        return result

    series, parallel = parsed
    views = result.setdefault("views", {})
    if "pack" in views:
        _adapt_pack_view(views["pack"], pack, series, parallel)
    if "bms" in views:
        components = result.setdefault("components", {})
        _adapt_bms_view(views["bms"], pack, series, components)

    return result

File: apps/web/src/lib/bms/topologyLayout.ts
import type { BmsArchitecture, PackInfo, SchematicEdge, SchematicNode, SchematicView } from "./types";
import { NODE_ROLE_LABELS } from "./blockLabel";
import { parsePackTopology, resolveComponentPart, seriesCellPinLabel } from "./layout";

export type PackTopology = NonNullable<ReturnType<typeof parsePackTopology>>;

/** Cell-stack block size scales with series × parallel so 4S and 12S are visibly different. */
export function cellStackDimensions(series: number, parallel: number): { width: number; height: number } {
  return {
    width: Math.min(720, Math.max(140, 48 + series * 22)),
    height: Math.max(72, 56 + Math.max(0, parallel - 1) * 36),
  };
}

/** AFE monitor block grows with series count. */
export function afeBlockDimensions(series: number): { width: number; height: number } {
  return {
    width: Math.min(300, Math.max(170, 100 + series * 7)),
    height: Math.max(95, 80 + Math.floor(series / 4) * 10),
  };
}

function ensureBlockSize(node: SchematicNode, minWidth: number, minHeight: number): void {
  node.width = Math.max(node.width, minWidth);
  node.height = Math.max(node.height, minHeight);
}

function cloneNode(node: SchematicNode): SchematicNode {
  return {
    ...node,
    pins: node.pins?.map((p) => ({ ...p })),
    telemetry: node.telemetry ? { ...node.telemetry } : undefined,
  };
}

function cloneView(view: SchematicView): SchematicView {
  return {
    nodes: view.nodes.map(cloneNode),
    edges: view.edges.map((e) => ({ ...e })),
  };
}

function findNode(nodes: SchematicNode[], id: string): SchematicNode | undefined {
  return nodes.find((n) => n.id === id);
}

function findByType(nodes: SchematicNode[], type: string): SchematicNode | undefined {
  return nodes.find((n) => n.type === type);
}

function updatePinLabel(node: SchematicNode, prefix: RegExp, label: string): void {
  for (const pin of node.pins ?? []) {
    if (prefix.test(pin.label)) {
      pin.label = label;
    }
  }
}

function formatStackLabel(topology: PackTopology, pack: PackInfo): string {
  const topo = `${topology.series}S${topology.parallel > 1 ? `${topology.parallel}P` : ""}`;
  return `${topo} ${pack.chemistry} Stack`;
}

function shiftNodesRight(nodes: SchematicNode[], fromX: number, deltaX: number): void {
  if (deltaX <= 0) return;
  for (const node of nodes) {
    if (node.x >= fromX) {
      node.x += deltaX;
    }
  }
}

export function adaptPackView(view: SchematicView, pack: PackInfo, topology: PackTopology): SchematicView {
  const adapted = cloneView(view);
  const cells = findByType(adapted.nodes, "cells");
  if (!cells) return adapted;

  const prevWidth = cells.width;
  const dims = cellStackDimensions(topology.series, topology.parallel);
  cells.width = dims.width;
  cells.height = dims.height;
  cells.label = formatStackLabel(topology, pack);
  cells.y = Math.max(280, 460 - dims.height);

  const widthDelta = cells.width - prevWidth;
  if (widthDelta > 0) {
    shiftNodesRight(adapted.nodes, cells.x + prevWidth + 20, widthDelta);
  }

  const bmsBoard = findByType(adapted.nodes, "bms_board");
  if (bmsBoard) {
    const minGap = 44;
    bmsBoard.x = Math.max(bmsBoard.x, cells.x + cells.width + minGap);
  }

  const contactor = findByType(adapted.nodes, "contactor");
  if (contactor && bmsBoard) {
    contactor.x = Math.max(contactor.x, bmsBoard.x + bmsBoard.width + 36);
  }

  return adapted;
}

function injectSecondAfe(
  nodes: SchematicNode[],
  edges: SchematicEdge[],
  primary: SchematicNode,
  series: number,
  part: string | null,
): void {
  if (series <= 16 || nodes.some((n) => n.id === "cell_ic_2")) return;

  const secondSeries = series - 16;
  const startCell = 17;
  const endCell = startCell + secondSeries - 1;
  const cellLabel =
    secondSeries <= 1 ? `CELL_${startCell}` : `CELL_${startCell}-${endCell}`;

  const yOffset = primary.height + 36;
  nodes.push({
    id: "cell_ic_2",
    type: "cell_monitor_ic",
    label: part ? `${part} AFE (B)` : "Cell Monitor IC (B)",
    x: primary.x,
    y: primary.y + yOffset,
    width: afeBlockDimensions(secondSeries).width,
    height: afeBlockDimensions(secondSeries).height,
    component_ref: primary.component_ref ? `${primary.component_ref}_b` : undefined,
    telemetry: { cells_monitored: secondSeries },
    pins: [
      { id: "cell_sense", label: cellLabel, side: "left" },
      { id: "i2c_sda", label: "I2C_SDA", side: "top" },
      { id: "i2c_scl", label: "I2C_SCL", side: "top" },
      { id: "bal_out", label: "BAL_OUT", side: "right" },
    ],
  });

  const mcu = findByType(nodes, "mcu");
  if (mcu) {
    edges.push({
      from_node: mcu.id,
      from_pin: "i2c_sda",
      to_node: "cell_ic_2",
      to_pin: "i2c_sda",
      signal: "I2C_SDA",
    });
    edges.push({
      from_node: mcu.id,
      from_pin: "i2c_scl",
      to_node: "cell_ic_2",
      to_pin: "i2c_scl",
      signal: "I2C_SCL",
    });
  }
}

function enrichNodeLabel(
  node: SchematicNode,
  components?: Record<string, Record<string, unknown>>,
): void {
  const part = resolveComponentPart(node.component_ref, components);
  if (!part) return;
  if (/\s[–—-]\s/.test(node.label)) return;

  const role = NODE_ROLE_LABELS[node.type] ?? node.type.replace(/_/g, " ");
  if (node.type === "cells" || node.type === "bms_board" || node.type === "contactor") return;
  node.label = `${role} – ${part}`;
}

function resolvePartLabel(node: SchematicNode, components?: Record<string, Record<string, unknown>>): void {
  enrichNodeLabel(node, components);
}

export function adaptBmsView(
  view: SchematicView,
  pack: PackInfo,
  topology: PackTopology,
  components?: Record<string, Record<string, unknown>>,
): SchematicView {
  const adapted = cloneView(view);
  const { series } = topology;
  const cellLabel = seriesCellPinLabel(series);
  const ntcLabel = series <= 1 ? "NTC_1" : `NTC_1-${series}`;

  const cellIc = findByType(adapted.nodes, "cell_monitor_ic");
  if (cellIc) {
    const primarySeries = series > 16 ? 16 : series;
    const afeDims = afeBlockDimensions(primarySeries);
    cellIc.width = afeDims.width;
    cellIc.height = afeDims.height;
    updatePinLabel(cellIc, /^CELL_/i, seriesCellPinLabel(primarySeries));
    cellIc.telemetry = { ...cellIc.telemetry, cells_monitored: primarySeries };
    resolvePartLabel(cellIc, components);

    const part =
      cellIc.component_ref && components?.[cellIc.component_ref]?.part;
    const partStr = typeof part === "string" ? part : null;
    if (series > 16) {
      injectSecondAfe(adapted.nodes, adapted.edges, cellIc, series, partStr);
    } else {
      updatePinLabel(cellIc, /^CELL_/i, cellLabel);
      cellIc.telemetry = { ...cellIc.telemetry, cells_monitored: series };
    }
  }

  const tempNet = findByType(adapted.nodes, "temperature_network");
  if (tempNet) {
    updatePinLabel(tempNet, /^NTC_/i, ntcLabel);
    if (cellIc) {
      tempNet.y = Math.max(tempNet.y, cellIc.y + cellIc.height + 28);
    }
  }

  const balancing = findByType(adapted.nodes, "balancing");
  if (balancing && cellIc) {
    balancing.y = Math.max(balancing.y, cellIc.y + cellIc.height + 20);
  }

  const current = findByType(adapted.nodes, "current_sensor");
  const memory = findByType(adapted.nodes, "memory");
  if (current && memory) {
    const bottomY = Math.max(current.y, memory.y, 320);
    current.y = bottomY;
    memory.y = bottomY + current.height + 24;
    if (current.x >= memory.x - 40 && current.x <= memory.x + memory.width) {
      memory.x = current.x + current.width + 40;
    }
  }

  for (const node of adapted.nodes) {
    resolvePartLabel(node, components);
    if (node.type === "mcu") ensureBlockSize(node, 160, 100);
    if (node.type === "can_transceiver") ensureBlockSize(node, 150, 85);
    if (node.type === "current_sensor" || node.type === "memory") ensureBlockSize(node, 150, 88);
    if (node.type === "temperature_network" || node.type === "balancing") ensureBlockSize(node, 150, 85);
  }

  if (components?.temp_network?.sensor_count !== series) {
    // display-only hint via telemetry on temp node
    if (tempNet) {
      tempNet.telemetry = { ...tempNet.telemetry, sensor_count: series };
    }
  }

  return adapted;
}

/** Reshape schematic layout from pack topology (display + save normalization). */
export function adaptArchitectureForTopology(architecture: BmsArchitecture): BmsArchitecture {
  const topology = parsePackTopology(architecture.pack.topology);
  if (!topology) return architecture;

  return {
    ...architecture,
    views: {
      pack: adaptPackView(architecture.views.pack, architecture.pack, topology),
      bms: adaptBmsView(architecture.views.bms, architecture.pack, topology, architecture.components),
    },
  };
}
